src/server/src/main/java/io/cassandrareaper/service/SegmentGenerator.java

```diff
@@ -81,76 +87,186 @@ static boolean greaterThanOrEqual(BigInteger big0, BigInteger big1) {
-        BigInteger offset = rs.multiply(BigInteger.valueOf(j)).divide(BigInteger.valueOf(segmentCount));
-        BigInteger reaperToken = start.add(offset);
-        if (greaterThan(reaperToken, rangeMax)) {
-          reaperToken = reaperToken.subtract(rangeSize);
+        // Append the segments between the endpoints
+        for (int j = 0; j < segmentCount; j++) {
+          repairSegments.add(
+              Segment.builder()
+                  .withTokenRanges(
+                      Arrays.asList(
+                          new RingRange(endpointTokens.get(j), endpointTokens.get(j + 1))))
+                  .build());
+          LOG.debug(
+              "Segment #{}: [{},{})", j + 1, endpointTokens.get(j), endpointTokens.get(j + 1));
         }
-        endpointTokens.add(reaperToken);
       }
 
-      // Append the segments between the endpoints
-      for (int j = 0; j < segmentCount; j++) {
-        repairSegments.add(new RingRange(endpointTokens.get(j), endpointTokens.get(j + 1)));
-        LOG.debug("Segment #{}: [{},{})", j + 1, endpointTokens.get(j), endpointTokens.get(j + 1));
+      // verify that the whole range is repaired
+      BigInteger total = BigInteger.ZERO;
+      for (Segment segment : repairSegments) {
+        for (RingRange range : segment.getTokenRanges()) {
+          BigInteger size = range.span(rangeSize);
+          total = total.add(size);
+        }
       }
+      if (!total.equals(rangeSize) && !incrementalRepair) {
+        throw new ReaperException("Not entire ring would get repaired");
+      }
+    } else {
+      // We want less segments than there are token ranges.
+      // Token ranges will be grouped to match the requirements.
+      LOG.info("Less segments required than there are vnode. Coalescing eligible token ranges...");
+      repairSegments =
+          coalesceTokenRanges(getTargetSegmentSize(totalSegmentCount), replicasToRange);
     }
 
-    // verify that the whole range is repaired
-    BigInteger total = BigInteger.ZERO;
-    for (RingRange segment : repairSegments) {
-      BigInteger size = segment.span(rangeSize);
-      total = total.add(size);
+    return repairSegments;
+  }
+
+  @VisibleForTesting
+  List<Segment> coalesceTokenRanges(
+      BigInteger targetSegmentSize, Map<List<String>, List<RingRange>> replicasToRange) {
+
+    List<Segment> coalescedRepairSegments = Lists.newArrayList();
+    List<RingRange> tokenRangesForCurrentSegment = Lists.newArrayList();
+    BigInteger tokenCount = BigInteger.ZERO;
+
+    for (Entry<List<String>, List<RingRange>> tokenRangesByReplica : replicasToRange.entrySet()) {
+      LOG.info("Coalescing segments for nodes {}", tokenRangesByReplica.getKey());
+      if (!tokenRangesForCurrentSegment.isEmpty()) {
+        coalescedRepairSegments.add(
+            Segment.builder().withTokenRanges(tokenRangesForCurrentSegment).build());
+        tokenRangesForCurrentSegment = Lists.newArrayList();
+      }
```

Comment:
i would expect lines 201-205 to go at the end of the for-loop (ie line 223), so that you could remove lines 227-230.

Reply:
right, good catch !